From 635f9e92373ca2deecd85f5a8d31eefb8d25d8ac Mon Sep 17 00:00:00 2001
From: Jiafei Pan <Jiafei.Pan@nxp.com>
Date: Fri, 13 Mar 2026 14:14:24 +0800
Subject: [PATCH 01/13] feat(gicv3): add APIs for clearing pending interrupts

Add two new APIs:
1. Clean interrupts from distributor for a CPU Core, it will clear
   pending SPIs interrupts and disable all interrups routes to this
   CPU Core.
2. Clean interrupts from redistributor interface, it will clean
   pending and disable all SGIs/PPIs on this redistributor interface.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
Upstream-Status: Submitted
---
 drivers/arm/gic/v3/gicv3_helpers.c | 40 ++++++++++++++++++++++++++++++
 drivers/arm/gic/v3/gicv3_main.c    | 28 +++++++++++++++++++++
 drivers/arm/gic/v3/gicv3_private.h |  2 ++
 include/drivers/arm/gicv3.h        |  2 ++
 4 files changed, 72 insertions(+)

diff --git a/drivers/arm/gic/v3/gicv3_helpers.c b/drivers/arm/gic/v3/gicv3_helpers.c
index d0a84605f..dccb99495 100644
--- a/drivers/arm/gic/v3/gicv3_helpers.c
+++ b/drivers/arm/gic/v3/gicv3_helpers.c
@@ -323,6 +323,46 @@ void gicv3_clear_spi_cluster(uintptr_t gicd_base)
 	}
 }
 
+/*******************************************************************************
+ * Helper function to clear all SPI interrupts of a CPU Core
+ ******************************************************************************/
+void gicv3_clear_spi_cpu(uintptr_t gicd_base)
+{
+	unsigned int irq_num, num_ints;
+	unsigned long long cpu_aff, irq_aff;	/* the irouter affinity values */
+	unsigned int mpidr = read_mpidr();
+
+	num_ints = gicd_read_typer(gicd_base);
+	num_ints &= TYPER_IT_LINES_NO_MASK;
+	num_ints = (num_ints + 1) << 5;
+
+	cpu_aff = gicd_irouter_val_from_mpidr(mpidr, 0);
+
+	/* browse all SPIs */
+	for (irq_num = MIN_SPI_ID; irq_num < num_ints; irq_num++) {
+		irq_aff = gicd_read_irouter(gicd_base, irq_num);
+		if ((irq_aff & MPIDR_AFFINITY_MASK) == cpu_aff) {
+			/* if affinity matches CPU Core, clear INT */
+			/* disable INT */
+			gicd_set_icenabler(gicd_base, irq_num);
+			/* remove pending state */
+			gicd_set_icpendr(gicd_base, irq_num);
+			while (gicd_read_ctlr(gicd_base) & GICD_CTLR_RWP_BIT);
+		}
+	}
+}
+
+/*******************************************************************************
+ * Helper function to clear all SGI/PPI interrupts of a GIC redistributor
+ ******************************************************************************/
+void gicv3_clear_sgi_ppi_cpu(uintptr_t gicr_base)
+{
+	/* Disable SGIs: 0-15, PPIs: 16-31 */
+	gicr_write_isenabler0(gicr_base, 0);
+	/* Clean Pending SGIs and PPIs */
+	gicr_write_icpendr0(gicr_base, 0);
+}
+
 /*******************************************************************************
  * Helper function to configure the default attributes of (E)PPIs/SGIs
  ******************************************************************************/
diff --git a/drivers/arm/gic/v3/gicv3_main.c b/drivers/arm/gic/v3/gicv3_main.c
index 81c4f6545..a890a8d6c 100644
--- a/drivers/arm/gic/v3/gicv3_main.c
+++ b/drivers/arm/gic/v3/gicv3_main.c
@@ -230,6 +230,34 @@ void gicv3_distif_clear(void)
 	gicv3_clear_spi_cluster(gicv3_driver_data->gicd_base);
 }
 
+/*******************************************************************************
+ * This function clears interrupts for CPU Core at GIC distributor level
+ ******************************************************************************/
+void gicv3_distif_cpu_clear(void)
+{
+	assert(gicv3_driver_data);
+	assert(gicv3_driver_data->gicd_base);
+
+	gicv3_clear_spi_cpu(gicv3_driver_data->gicd_base);
+}
+
+/*******************************************************************************
+ * This function clears interrupts for CPU Core at GIC redistributor level
+ ******************************************************************************/
+void gicv3_rdistif_cpu_clear(unsigned int proc_num)
+{
+	uintptr_t gicr_base;
+
+	assert(gicv3_driver_data);
+	assert(proc_num < gicv3_driver_data->rdistif_num);
+	assert(gicv3_driver_data->rdistif_base_addrs != NULL);
+
+	gicr_base = gicv3_driver_data->rdistif_base_addrs[proc_num];
+	assert(gicr_base != 0UL);
+
+	gicv3_clear_sgi_ppi_cpu(gicr_base);
+}
+
 /*******************************************************************************
  * This function initialises the GIC Redistributor interface of the calling CPU
  * (identified by the 'proc_num' parameter) based upon the data provided by the
diff --git a/drivers/arm/gic/v3/gicv3_private.h b/drivers/arm/gic/v3/gicv3_private.h
index ac7c5199f..b96d499eb 100644
--- a/drivers/arm/gic/v3/gicv3_private.h
+++ b/drivers/arm/gic/v3/gicv3_private.h
@@ -252,6 +252,8 @@ void gicv3_rdistif_base_addrs_probe(uintptr_t *rdistif_base_addrs,
 					mpidr_hash_fn mpidr_to_core_pos);
 void gicv3_rdistif_mark_core_awake(uintptr_t gicr_base);
 void gicv3_rdistif_mark_core_asleep(uintptr_t gicr_base);
+void gicv3_clear_spi_cpu(uintptr_t gicd_base);
+void gicv3_clear_sgi_ppi_cpu(uintptr_t gicr_base);
 
 /*******************************************************************************
  * GIC Distributor interface accessors
diff --git a/include/drivers/arm/gicv3.h b/include/drivers/arm/gicv3.h
index ebd39999a..5d1a121e9 100644
--- a/include/drivers/arm/gicv3.h
+++ b/include/drivers/arm/gicv3.h
@@ -603,6 +603,8 @@ static inline void gicv3_apply_errata_wa_2384374(const uintptr_t gicr_base)
 {
 }
 #endif /* GIC600_ERRATA_WA_2384374 */
+void gicv3_distif_cpu_clear(void);
+void gicv3_rdistif_cpu_clear(unsigned int proc_num);
 
 #endif /* __ASSEMBLER__ */
 #endif /* GICV3_H */
-- 
2.43.0

